### ISSUE TITLE
definition of "server"

### ISSUE THREAD

**Issue by jogu** [danielfett/draft-dpop#117]
I'm unsure exactly how to interpret this MUST:

https://github.com/danielfett/draft-dpop/blame/master/main.md#L1018-L1021

```
Note that the nonces provided by the two kinds of servers are different
and MUST not be confused with one another.
In particular, a nonce provided to the client by a particular server
MUST only be used with that server and no other.
```

I think `the two kinds of servers` partly confuses me (if an authorization server is hosting a resource endpoint like the user info endpoint, is it still considered an authorization server in this clause when calling userinfo?).

I'm then not sure on the meaning of 'server' - I guess given there may be multiple resource servers in play, it might mean something like "endpoints with the same scheme/hostname/portnumber" (i.e. comparing urls but ignoring path, query and fragment).

Does the 'MUST' imply there's a security implication if the client sent the authorization server's nonce to a resource endpoint?

**Comment by bc-pi** [comment]
I can see the potential for confusion there, yeah. 

> is it still considered an authorization server in this clause when calling userinfo?

yes

> Does the 'MUST' imply there's a security implication if the client sent the authorization server's nonce to a resource endpoint?

I don't think so. I belive it's just an overzealous keyword.

**Comment by b---c** [comment]
> I'm then not sure on the meaning of 'server' - I guess given there may be multiple resource servers in play, it might mean something like "endpoints with the same scheme/hostname/portnumber" (i.e. comparing urls but ignoring path, query and fragment).

I think the intent is (or should be) that server means domain. A client sends a nonce back to the same domain. The wording is awkward for me though.

**Comment by jogu** [comment]
> I think the intent is (or should be) that server means domain. A client sends a nonce back to the same domain. The wording is awkward for me though.

Thanks Brian. To check if I understand, you're saying that if I had an access token that gives me access to both of (say) email-api.microsoft.com and drive-api.microsoft.com, those two are both in the 'microsoft.com' domain and hence should be colluding on the nonce they expect?

(This seems to be mostly a performance concern, I think if the assumption was incorrect the client would just get more rejections & retries than it ideally would.)

**Comment by selfissued** [comment]
The two kinds of servers being referred to are authorization servers and resource servers.  I'll create a PR to clarify.

**Comment by bc-pi** [comment]
> Thanks Brian. To check if I understand, you're saying that if I had an access token that gives me access to both of (say) email-api.microsoft.com and drive-api.microsoft.com, those two are both in the 'microsoft.com' domain and hence should be colluding on the nonce they expect?

No, fully qualified domain. Or host. Or authority. Not sure the proper name :/  Each of those three you mention would be different. 

> (This seems to be mostly a performance concern, I think if the assumption was incorrect the client would just get more rejections & retries than it ideally would.)

Yeah, I do think that's the case. And is noted above in the AS nonce section.

**Comment by b---c** [comment]
> The two kinds of servers being referred to are authorization servers and resource servers. I'll create a PR to clarify.

I think that clarifying the scope of the nonce would be worthwhile too.
